File: apps/desktop/src/services/drafts.ts
/**
 * What somebody had started typing, per channel, kept until they send it.
 *
 * Switching away from a half-written message and back used to hand back an
 * empty box, and a restart did the same to everything. Every chat app keeps
 * these; the question was only where.
 *
 * - **In memory first.** The box writes here on every keystroke and reads from
 *   here on every mount, synchronously, so opening a channel never shows an
 *   empty composer that fills a frame later.
 * - **On disk behind it**, in the same IndexedDB store as the message cache
 *   (`services/cache.ts`), debounced: a write per keystroke is a write nobody
 *   needs, and the last half second of typing is flushed when the page goes.
 *   Living in that store is what makes a sign-out or an account switch take
 *   the drafts with it - `cache.clear()` is already the one door out.
 * - **Never to the server.** A draft is the one piece of plaintext this device
 *   keeps, and it is the person's own words on their own machine - the same
 *   trust as the text sitting in the box. Nothing here imports `api`.
 *
 * Every storage failure is swallowed: a draft that did not persist is a draft
 * lost on restart, and a composer that throws is a composer nobody can use.
 */
import { useEffect, useSyncExternalStore } from 'react';
import type { MessageReply } from '@betweenus/shared-types';
import { cache } from './cache';

export interface Draft {
  /** Exactly what was in the box, whitespace and all - it is theirs to trim. */
  text: string;
  /** The message being answered, if the box had one. */
  replyTo: MessageReply | null;
}

/** How long typing has to pause before the draft is written down. */
export const DRAFT_WRITE_DELAY_MS = 500;

/**
 * A draft worth keeping, or `null` for one that is nothing.
 *
 * Whitespace alone is nothing - a box somebody pressed space in is not a
 * draft, and a "Draft" label beside it in the sidebar would be a lie. A reply
 * with no text yet is something: choosing "Reply" is the start of a message.
 */
export function keptDraft(text: string, replyTo: MessageReply | null): Draft | null {
  if (text.trim().length === 0 && !replyTo) return null;
  return { text, replyTo };
}

const drafts = new Map<string, Draft>();
const listeners = new Set<() => void>();
/**
 * Channels whose box has been written to this session. The disk's copy of
 * those is older than memory's - including when memory's is "nothing, it was
 * sent" - so the first read off disk leaves them alone.
 */
const touched = new Set<string>();
/** Bumped on every publish, as the snapshot `useSyncExternalStore` compares. */
let version = 0;
let loading: Promise<void> | null = null;
let pending: ReturnType<typeof setTimeout> | null = null;
/**
 * Bumped by `forgetDrafts`, so a disk read that was already in flight when
 * somebody signed out cannot land afterwards and bring their drafts back.
 */
let generation = 0;

function publish(): void {
  version += 1;
  for (const listener of listeners) listener();
}

/** Called whenever the set of channels with a draft changes - not on every keystroke. */
export function onDraftsChanged(listener: () => void): () => void {
  listeners.add(listener);
  return () => listeners.delete(listener);
}

/** What was left in a channel's box, if anything. */
export function draftFor(channelId: string): Draft | null {
  return drafts.get(channelId) ?? null;
}

export function hasDraft(channelId: string): boolean {
  return drafts.has(channelId);
}

/**
 * `hasDraft`, for a list: the component re-renders when any channel gains or
 * loses a draft, and asks per row with what this returns.
 *
 * Starts the disk read itself: the list is on screen before any composer is,
 * and after a restart it is the list that has to remember.
 */
export function useDrafted(): (channelId: string) => boolean {
  useEffect(() => {
    void loadDrafts();
  }, []);
  useSyncExternalStore(onDraftsChanged, () => version);
  return hasDraft;
}

/**
 * Reads the drafts off disk, once.
 *
 * Merged under memory rather than over it: a box typed into (or sent from)
 * before the read came back holds something newer than the disk does.
 */
export function loadDrafts(): Promise<void> {
  if (loading) return loading;
  const reading = generation;
  loading = cache
    .drafts()
    .then((stored) => {
      if (reading !== generation || !stored) return;
      let changed = false;
      for (const [channelId, draft] of Object.entries(stored)) {
        const kept = keptDraft(draft.text, draft.replyTo);
        if (kept && !touched.has(channelId)) {
          drafts.set(channelId, kept);
          changed = true;
        }
      }
      if (changed) publish();
    })
    .catch(() => undefined);
  return loading;
}

/**
 * Remembers what is in a channel's box. An empty box forgets it.
 *
 * Memory now, disk after the pause. Only a draft appearing or going wakes the
 * listeners, so a sidebar does not repaint on every letter.
 */
export function saveDraft(channelId: string, text: string, replyTo: MessageReply | null): void {
  const kept = keptDraft(text, replyTo);
  touched.add(channelId);
  const had = drafts.has(channelId);
  const current = drafts.get(channelId);
  if (!kept && !had) return;
  if (kept && current && current.text === kept.text && current.replyTo?.id === kept.replyTo?.id) {
    return;
  }

  if (kept) drafts.set(channelId, kept);
  else drafts.delete(channelId);
  if (had !== Boolean(kept)) publish();
  schedule();
}

/** A message went: its draft goes now, not after the pause. */
export function clearDraft(channelId: string): void {
  touched.add(channelId);
  if (!drafts.delete(channelId)) return;
  publish();
  void flushDrafts();
}

/**
 * Forgets every draft in memory and drops the write that was waiting.
 *
 * For sign-out, beside `cache.clear()` - which empties the disk, and which a
 * debounced write landing half a second later would quietly refill.
 */
export function forgetDrafts(): void {
  if (pending) clearTimeout(pending);
  pending = null;
  generation += 1;
  loading = null;
  touched.clear();
  const had = drafts.size > 0;
  drafts.clear();
  if (had) publish();
}

function schedule(): void {
  if (pending) clearTimeout(pending);
  pending = setTimeout(() => void flushDrafts(), DRAFT_WRITE_DELAY_MS);
}

/** Writes the drafts now, if a write is waiting. */
export async function flushDrafts(): Promise<void> {
  if (pending) clearTimeout(pending);
  pending = null;
  const writing = generation;
  // Only once the disk has been read: writing the in-memory map before then
  // would replace every draft from last session with this session's few.
  await loadDrafts();
  if (writing !== generation) return;
  await cache.putDrafts(Object.fromEntries(drafts)).catch(() => undefined);
}

// The last half second of typing before the window closes. Best effort: the
// write is started, and IndexedDB usually finishes what it has started.
if (typeof window !== 'undefined') {
  window.addEventListener('pagehide', () => {
    if (pending) void flushDrafts();
  });
}

File: apps/android/core/src/main/java/com/aatech/betweenus/core/store/Drafts.kt
package com.aatech.betweenus.core.store

import com.aatech.betweenus.core.data.MessageReply
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONObject

/** What was left in a conversation's box: the text exactly as typed, and the reply if there was one. */
data class Draft(val text: String, val replyTo: MessageReply?)

/**
 * What somebody had started typing, per conversation, kept until they send it.
 *
 * The port of `apps/desktop/src/services/drafts.ts`, and the same three rules:
 *
 * - **Memory first.** The composer writes here on every keystroke and reads
 *   from here when it opens, so a conversation never opens on an empty box
 *   that fills a moment later.
 * - **Disk behind it**, as one row in [Cache] - debounced, because a Room
 *   write per letter is a write nobody needs. Living there is what makes a
 *   sign-out or a different account take the drafts with it: [Cache.clear] and
 *   [Cache.claim] are already the doors out.
 * - **Never to the server.** A draft is the one piece of plaintext this phone
 *   keeps, and it is the person's own words on their own device - the same
 *   trust as the text sitting in the box. Nothing here talks to the network.
 *
 * Every storage failure is swallowed: a draft that did not persist is one lost
 * on restart, and a composer that throws is one nobody can type into.
 */
object Drafts {
    /** How long typing has to pause before the draft is written down. */
    const val WRITE_DELAY_MS = 500L

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val lock = Any()
    private val loadLock = Mutex()

    private val drafts = mutableMapOf<String, Draft>()
    /**
     * Conversations whose box has been written to since the last read off
     * disk. Disk's copy of those is older than memory's - including when
     * memory's is "nothing, it was sent" - so the read leaves them alone.
     */
    private val touched = mutableSetOf<String>()

    @Volatile
    private var loaded = false
    private var pending: Job? = null

    /**
     * The conversations with something unsent in them, for the drawer's label.
     * Changes when a draft appears or goes, not on every letter.
     */
    private val _drafted = MutableStateFlow<Set<String>>(emptySet())
    val drafted: StateFlow<Set<String>> = _drafted.asStateFlow()

    /** What was left in a conversation's box, if anything. */
    fun draftFor(channelId: String): Draft? = synchronized(lock) { drafts[channelId] }

    /**
     * A draft worth keeping, or null for one that is nothing.
     *
     * Whitespace alone is nothing - a label saying "Draft" beside a box
     * somebody pressed space in would be a lie. A reply with no words yet is
     * something: choosing Reply is the start of a message.
     */
    internal fun kept(text: String, replyTo: MessageReply?): Draft? =
        if (text.isBlank() && replyTo == null) null else Draft(text, replyTo)

    /**
     * Reads the drafts off disk, once. Safe to call from every screen that
     * wants them; the second call is free.
     */
    suspend fun load() {
        if (loaded) return
        loadLock.withLock {
            if (loaded) return
            val stored = runCatching { Cache.drafts() }.getOrNull()?.let(::decode).orEmpty()
            synchronized(lock) {
                stored.forEach { (channelId, draft) ->
                    if (channelId !in touched) drafts[channelId] = draft
                }
                loaded = true
                publish()
            }
        }
    }

    /**
     * Remembers what is in a conversation's box. An empty box forgets it.
     * Memory now, disk after the pause.
     */
    fun save(channelId: String, text: String, replyTo: MessageReply?) {
        val next = kept(text, replyTo)
        synchronized(lock) {
            touched += channelId
            if (drafts[channelId] == next) return
            if (next == null) drafts.remove(channelId) else drafts[channelId] = next
            publish()
        }
        schedule(delayMs = WRITE_DELAY_MS)
    }

    /** A message went: its draft goes now, not after the pause. */
    fun clear(channelId: String) {
        synchronized(lock) {
            touched += channelId
            if (drafts.remove(channelId) == null) return
            publish()
        }
        schedule(delayMs = 0)
    }

    /**
     * Forgets every draft and drops the write that was waiting.
     *
     * For sign-out, before [Cache.clear]: that empties the disk, and a
     * debounced write landing half a second later would quietly fill it again -
     * or, worse, wait for the next account's claim and land in theirs.
     */
    fun forget() {
        synchronized(lock) {
            pending?.cancel()
            pending = null
            drafts.clear()
            touched.clear()
            loaded = false
            publish()
        }
    }

    private fun publish() {
        _drafted.value = drafts.keys.toSet()
    }

    private fun schedule(delayMs: Long) {
        synchronized(lock) {
            pending?.cancel()
            pending = scope.launch {
                delay(delayMs)
                // Only once the disk has been read: writing memory's map before
                // then would replace last session's drafts with this one's few.
                load()
                val snapshot = synchronized(lock) { drafts.toMap() }
                runCatching { Cache.putDrafts(encode(snapshot)) }
            }
        }
    }

    internal fun encode(drafts: Map<String, Draft>): String {
        val json = JSONObject()
        drafts.forEach { (channelId, draft) ->
            json.put(
                channelId,
                JSONObject()
                    .put("text", draft.text)
                    .put("replyTo", draft.replyTo?.toJson() ?: JSONObject.NULL),
            )
        }
        return json.toString()
    }

    /** Anything unreadable, or anything that is nothing, is dropped rather than thrown. */
    internal fun decode(stored: String): Map<String, Draft> = runCatching {
        val json = JSONObject(stored)
        json.keys().asSequence().mapNotNull { channelId ->
            val row = json.optJSONObject(channelId) ?: return@mapNotNull null
            val reply = row.optJSONObject("replyTo")?.let { MessageReply.from(it) }
            kept(row.optString("text"), reply)?.let { channelId to it }
        }.toMap()
    }.getOrDefault(emptyMap())
}
